File: solutions/cpp/Deletion_list.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Linked_list_prototype.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::shared_ptr;

// @include
void DeletionFromList(const shared_ptr<ListNode<int>>& v) {
  v->data = v->next->data;
  v->next = v->next->next;
}
// @exclude

int main(int argc, char* argv[]) {
  shared_ptr<ListNode<int>> L;
  L = make_shared<ListNode<int>>(ListNode<int>{
      1, make_shared<ListNode<int>>(ListNode<int>{
             2, make_shared<ListNode<int>>(ListNode<int>{3, nullptr})})});
  DeletionFromList(L);
  assert(L->data == 2 && L->next->data == 3);
  return 0;
}

File: static/solutions/cpp/java/src/main/java/com/epi/DeletionList.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
// @author Andrey Pavlov

package com.epi;

public class DeletionList {
  // @include
  // Assumes nodeToDelete is not tail.
  public static void deletionFromList(ListNode<Integer> nodeToDelete) {
    nodeToDelete.data = nodeToDelete.next.data;
    nodeToDelete.next = nodeToDelete.next.next;
  }
  // @exclude

  public static void main(String[] args) {
    ListNode<Integer> L
        = new ListNode<>(1, new ListNode<>(2, new ListNode<>(3, null)));
    deletionFromList(L);
    assert(L.data == 2 && L.next.data == 3);
  }
}
